src/signatures/commitment_and_public_key_signature.rs

```diff
@@ -0,0 +1,349 @@
+        };
+
+        // The challenge cannot be zero
+        if e == K::default() {
+            return Err(CommitmentAndPublicKeySignatureError::InvalidChallenge);
+        }
+
+        // Compute the response values
+        let ea = &e * a;
+        let ex = &e * x;
+        let ey = &e * y;
+
+        let u_a = r_a + &ea;
+        let u_x = r_x + &ex;
+        let u_y = r_y + &ey;
+
+        // Compute the initial values
+        let ephemeral_commitment = factory.commit(r_x, r_a);
+        let ephemeral_pubkey = P::from_secret_key(r_y);
+
+        Ok(Self::new(ephemeral_commitment, ephemeral_pubkey, u_a, u_x, u_y))
+    }
+
+    /// Verify a signature on a commitment and group element statement using a given challenge (as a byte array)
+    pub fn verify_challenge<'a, C>(
+        &self,
+        commitment: &'a HomomorphicCommitment<P>,
+        pubkey: &'a P,
+        challenge: &[u8],
+        factory: &C,
+    ) -> bool
+    where
+        for<'b> &'a HomomorphicCommitment<P>: Mul<&'b K, Output = HomomorphicCommitment<P>>,
+        for<'b> &'b P: Mul<&'b K, Output = P>,
+        for<'b> &'b HomomorphicCommitment<P>: Add<&'b HomomorphicCommitment<P>, Output = HomomorphicCommitment<P>>,
+        for<'b> &'b P: Add<&'b P, Output = P>,
+        for<'b> &'b K: Mul<&'b K, Output = K>,
+        for<'b> &'b K: Add<&'b K, Output = K>,
+        C: HomomorphicCommitmentFactory<P = P>,
+    {
+        // The challenge must be a valid scalar
+        let e = match K::from_bytes(challenge) {
+            Ok(e) => e,
+            Err(_) => return false,
+        };
+
+        self.verify(commitment, pubkey, &e, factory)
+    }
+
+    /// Verify a signature on a commitment and group element statement using a given challenge (as a scalar)
+    pub fn verify<'a, C>(&self, commitment: &'a HomomorphicCommitment<P>, pubkey: &'a P, challenge: &K, factory: &C) -> bool
+    where
+        for<'b> &'a HomomorphicCommitment<P>: Mul<&'b K, Output = HomomorphicCommitment<P>>,
+        for<'b> &'b P: Mul<&'b K, Output = P>,
+        for<'b> &'b HomomorphicCommitment<P>: Add<&'b HomomorphicCommitment<P>, Output = HomomorphicCommitment<P>>,
+        for<'b> &'b P: Add<&'b P, Output = P>,
+        for<'b> &'b K: Mul<&'b K, Output = K>,
+        for<'b> &'b K: Add<&'b K, Output = K>,
+        C: HomomorphicCommitmentFactory<P = P>,
+    {
+        // The challenge cannot be zero
+        if *challenge == K::default() {
+            return false
+        }
+
+        // Use a single weighted equation for verification to avoid unnecessary group operations
+        // For now, we use naive multiscalar multiplication, but offload the commitment computation
+        // This allows for the use of precomputation within the commitment itself, which is more efficient
+        // TODO: use curve library (Straus/Pippenger) multiscalar multiplication with partial precomputation
+        let mut rng = thread_rng();
```

Comment:
The RNG should be passed into the method as a parameter. Different platforms (e.g. WASM) may not have this RNG, so rather, add it as a parameter with the trait bound CryptoRng (I think this is the name)